### PR TITLE
fix(shell-wrapper): propagate login shell status

### PR DESCRIPTION
**Pull request by nzbr**
Starts the wrapped shell as a login shell if the shell wrapper is started as a login shell (arg0 is prefixed with `-`)

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Not working for me :/

```
May 06 16:55:15 francium systemd[1]: Stopping User Manager for UID 1000...
May 06 16:55:15 francium systemd[409]: Activating special unit Exit the Session...
May 06 16:55:15 francium ssh-agent[468]: exiting on signal 15
May 06 16:55:15 francium systemd[409]: Stopped target Main User Target.
May 06 16:55:15 francium systemd[409]: Stopping SSH Agent...
May 06 16:55:15 francium systemd[409]: Stopped SSH Agent.
May 06 16:55:15 francium systemd[409]: Stopped target Basic System.
May 06 16:55:15 francium systemd[409]: Stopped target Paths.
May 06 16:55:15 francium systemd[409]: Stopped target Sockets.
May 06 16:55:15 francium systemd[409]: Stopped target Timers.
May 06 16:55:15 francium systemd[409]: Closed D-Bus User Message Bus Socket.
May 06 16:55:15 francium systemd[409]: Closed GnuPG cryptographic agent and passphrase cache.
May 06 16:55:15 francium systemd[409]: Closed Socket for Lorri Daemon.
May 06 16:55:15 francium systemd[409]: Closed Podman API Socket.
May 06 16:55:15 francium systemd[409]: Removed slice User Application Slice.
May 06 16:55:15 francium systemd[409]: Reached target Shutdown.
May 06 16:55:15 francium systemd[409]: Finished Exit the Session.
May 06 16:55:15 francium systemd[409]: Reached target Exit the Session.
May 06 16:55:15 francium (sd-pam)[414]: pam_unix(systemd-user:session): session closed for user sandro
May 06 16:55:15 francium systemd[1]: user@1000.service: Deactivated successfully.
May 06 16:55:15 francium systemd[1]: Stopped User Manager for UID 1000.
```

**Comment by nzbr** [comment]
It does not fix systemd user sessions. Only that the shell is started as a login shell.
Can you see if systemd user sessions work with another distro on your system? Because they do not for me (tested Ubuntu and Debian so far)

**Comment by SuperSandro2000** [comment]
> Can you see if systemd user sessions work with another distro on your system?

They work when I go back to 0b90c1d982d443358b3f7b3a303405449a2bfe54. I don't think I enjoyed lingering or anything else. 

They also work since then but only for a few minutes and then they get closed.

**Comment by nzbr** [comment]
WSL starts a single shell under `login` to create the logind session. That shell runs in the background and is not the one that gets connected to the terminal. The process has SIGCHLD set to be ignored, which makes the shell wrapper crash when it tries to wait for the `sh` it forks of to source `/etc/set-environment`.
I have changed it so that it now checks if SIGCHLD is ignored and if yes it skips setting the environment.
tl;dr: it should work now

File: utils/src/shell_wrapper.rs [hunk 1]
```diff
@@ -18,37 +20,76 @@ fn real_main() -> anyhow::Result<()> {
...
+            unsafe {
+                // WSL starts a single shell under login to make sure that a logind session exists.
+                // That shell is started with SIGCHLD ignored
+                // If it is, we are probably that shell and can just skip setting the environment
+                let mut act: sigaction = MaybeUninit::zeroed().assume_init();
+                sigaction(SIGCHLD, PT_NULL as *const sigaction, &mut act);
+                if act.sa_sigaction == SIG_IGN {
+                    return Ok(());
+                }
+            }
```
Comment (K900) [inline utils/src/shell_wrapper.rs:38]: https://docs.rs/nix/latest/nix/sys/signal/fn.sigaction.html

File: utils/src/shell_wrapper.rs [hunk 2]
```diff
@@ -58,5 +99,12 @@ fn main() {
...
 
     env::set_var("RUST_BACKTRACE", "1");
-    eprintln!("[shell-wrapper] Error: {:?}", result.unwrap_err());
+    let err = result.unwrap_err();
+
+    eprintln!("[shell-wrapper] Error: {:?}", &err);
+
+    // Write the result to /tmp/shell-wrapper_crash_<pid>.log
+    let pid = std::process::id();
+    let output_file = format!("/tmp/shell-wrapper_crash_{}.log", pid);
```
Comment (K900) [inline utils/src/shell_wrapper.rs:108]: Can we log to the journal instead?

File: utils/src/shell_wrapper.rs (hunk 1 above)
Comment (K900) [reply utils/src/shell_wrapper.rs:38]: Never mind that can't read (by passing null), only write. Should probably add a comment for that.

Comment (nzbr) [reply utils/src/shell_wrapper.rs:42]: added a comment

File: utils/src/shell_wrapper.rs (hunk 2 above)
Comment (nzbr) [reply utils/src/shell_wrapper.rs]: It was a quick and dirty implementation so I could see what was going on. Journal sounds like a way better idea. I've changed it to do that instead
